README.md
# pwcm

A minimal working example of C++ function multiversioning in Python wheels.

* Date: 2017-02-18
* Version: 0.0.1
* Developer: [Alberto Pettarin](http://www.albertopettarin.it/)
* License: the MIT License (MIT)
* Contact: [click here](http://www.albertopettarin.it/contact.html)


## Overview

This repository contains a simple
Python 2.7/3.5+ package named ``pwcm``
that has a C++ extension (``pwcm.cext``) containing
[function multiversioning](https://gcc.gnu.org/wiki/FunctionMultiVersioning).
See also [this article](https://lwn.net/Articles/691932/).

The basic goal consists in testing the possibility of creating
a single Python wheel supporting CPUs with different features,
e.g. SSE3 vs SSE4.1/SSE4.2 vs AVX.
See this [PyTorch issue](https://github.com/pytorch/pytorch/issues/535).

So far, it seems to work as expected:

| Machine | CPU                        | Flags  | pwcm Output | Correct? | Compiler | Date       |
|---------|----------------------------|--------|-------------|----------|----------|------------|
| laptop1 | Intel(R) Core(TM) i5-6200U | SSE4.2 | SSE4.2      | Y        | GCC 6    | 2017-02-18 |
| vps1    | AMD Athlon(tm) II X2 240   | SSE4a  | SSE4a       | Y        | GCC 6    | 2017-02-18 |


## Installation from source

```bash
$ virtualenv venv
$ cd venv
$ source bin/activate
$ git clone https://github.com/pettarin/pwcm
$ cd pwcm
$ make test
$ deactivate
```

The last line of the output should be something like:

```plain
[INFO] C++ extension returned the known value: SSE4.2
```


## Create a Python wheel

```bash
$ # in the pwcm root directory, the one containing README.md
$ make wheel
$ # look inside the dist directory
$ # the name of the .whl file depends on your Python version
$ ls dist
pwcm-0.0.1-cp27-cp27mu-linux_x86_64.whl
```


## Installation from a wheel file

```bash
$ # create a new virtualenv
$ virtualenv venvtest
$ cd venvtest
$ source bin/activate
$ # download/copy the .whl file here
$ # adjust the name depending on your platform
$ pip install pwcm-0.0.1-cp27-cp27mu-linux_x86_64.whl
$ python -m pwcm
$ deactivate
```

Again, ``python -m pwcm`` should print something like:

```plain
[INFO] C++ extension returned the known value: SSE4.2
```


## Clean

```bash
$ # in the pwcm root directory, the one containing README.md
$ make clean
$ # in case you installed with pip
$ pip uninstall pwcm
```


## License

The contents of this repository are released
under the terms of the MIT License.


CHANGELOG
Changelog
=========

v0.0.1 (2017-02-18)
-------------------

#. Initial release


pwcm/cext.cpp
#include <Python.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cfunc.h"

// wrapper around _dosomething
static PyObject *dosomething(PyObject *self, PyObject *args) {
    return Py_BuildValue("i", _dosomething());
}

static PyMethodDef cext_methods[] = {
    {
        "dosomething",
        dosomething,
        METH_VARARGS,
        "Do something: returns a different int for each different version of the function"
    },
    {
        NULL,
        NULL,
        0,
        NULL
    }
};

// Python 2 and 3
#if PY_MAJOR_VERSION >= 3
static struct PyModuleDef moduledef = {
    PyModuleDef_HEAD_INIT,
    "cext",         /* m_name */
    "cext",         /* m_doc */
    -1,             /* m_size */
    cext_methods,   /* m_methods */
    NULL,           /* m_reload */
    NULL,           /* m_traverse */
    NULL,           /* m_clear */
    NULL,           /* m_free */
};
#endif

static PyObject *moduleinit(void) {
    PyObject *m;

#if PY_MAJOR_VERSION >= 3
    m = PyModule_Create(&moduledef);
#else
    m = Py_InitModule("cext", cext_methods);
#endif

    return m;
}

#if PY_MAJOR_VERSION >= 3
PyMODINIT_FUNC PyInit_cext(void) {
    return moduleinit();
}
#else
PyMODINIT_FUNC initcext(void) {
    moduleinit();
}
#endif


pwcm/__main__.py
#!/usr/bin/env python
# coding=utf-8

from __future__ import print_function
import sys

__author__ = "Alberto Pettarin"
__email__ = "alberto@albertopettarin.it"
__copyright__ = "Copyright 2017, Alberto Pettarin (www.albertopettarin.it)"
__license__ = "MIT"
__status__ = "Beta"
__version__ = "0.0.1"


MAP = {
    0: "default",
    1: "SSE1",
    2: "SSE2",
    3: "SSE3",
    40: "SSE4a",
    41: "SSE4.1",
    42: "SSE4.2"
}


def main():
    try:
        from pwcm.cext import dosomething
        v = dosomething()
        if v not in MAP:
            print("[ERRO] C++ extension returned the unknown value: %d" % v)
        else:
            print("[INFO] C++ extension returned the known value: %s" % MAP[v])
    except Exception as e:
        print("[ERRO] Unable to run C++ extension. Exception raised:")
        print(e)
    sys.exit(0)


if __name__ == "__main__":
    main()


setupmeta.py
#!/usr/bin/env python
# coding=utf-8

"""
Metadata for setting the pwcm package up
"""

import io

__author__ = "Alberto Pettarin"
__email__ = "alberto@albertopettarin.it"
__copyright__ = "Copyright 2017, Alberto Pettarin (www.albertopettarin.it)"
__license__ = "MIT"
__status__ = "Beta"
__version__ = "0.0.1"


##############################################################################
#
# you might need to edit the information below this line
#
##############################################################################

# package version
# NOTE: generate a new one for each PyPI upload, otherwise it will fail
PKG_VERSION = "0.0.1"

# required packages to install
# NOTE: always use exact version numbers
# NOTE: this list should be the same as requirements.txt
PKG_INSTALL_REQUIRES = []

# required packages to install extra tools
PKG_EXTRAS_REQUIRE = {}

# packages to be distributed
PKG_PACKAGES = [
    "pwcm",
]

# data files to be distributed
# NOTE: .py files will be added automatically
PKG_PACKAGE_DATA = {
    "pwcm": [
        "*.cpp",
        "*.h",
        "*.md"
    ],
}

# scripts to be installed globally
# on Linux and Mac OS X, use the file without extension
# on Windows, use the file with .py extension
PKG_SCRIPTS = []

##############################################################################
#
# do not edit the metadata below this line
#
##############################################################################

# package name
PKG_NAME = "pwcm"

# package author
PKG_AUTHOR = "Alberto Pettarin"

# package author email
PKG_AUTHOR_EMAIL = "alberto@albertopettarin.it"

# package URL
PKG_URL = "https://github.com/pettarin/pwcm"

# package license
PKG_LICENSE = "MIT"

# human-readable descriptions
PKG_SHORT_DESCRIPTION = """A minimal working example of
C++ function multiversioning in Python wheels."""
try:
    PKG_LONG_DESCRIPTION = io.open("README.rst", "r", encoding="utf-8").read()
except:
    PKG_LONG_DESCRIPTION = PKG_SHORT_DESCRIPTION

# PyPI keywords
PKG_KEYWORDS = [
    "C",
    "Python wheel",
    "Python",
    "function multiversioning",
    "function",
    "gcc",
    "multiversion function"
    "multiversion",
    "multiversioning",
    "wheel",
]

# PyPI classifiers
PKG_CLASSIFIERS = [
    "Development Status :: 4 - Beta",
    "Environment :: Console",
    "Intended Audience :: Developers",
    "Intended Audience :: Science/Research",
    "License :: OSI Approved :: MIT License",
    "Natural Language :: English",
    "Operating System :: MacOS :: MacOS X",
    "Operating System :: Microsoft :: Windows",
    "Operating System :: POSIX :: Linux",
    "Programming Language :: C",
    "Programming Language :: Python",
    "Programming Language :: Python :: 2",
    "Programming Language :: Python :: 2.7",
    "Programming Language :: Python :: 3",
    "Programming Language :: Python :: 3.5",
    "Topic :: Software Development :: Libraries :: Python Modules",
]
